### DISCUSSION TITLE
Stream read and stream response "hangs", returns 327 bytes of 10 megabytes

### DISCUSSION THREAD

**Discussion by voltagex** [fastapi/fastapi#11566] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.responses import Response, StreamingResponse
import logging
import aiohttp
import sys
logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)

app = FastAPI()

async def async_request():  # 
    async with aiohttp.ClientSession() as client:
        logging.info("Start get")
        r = client.get(url="https://mirror.aarnet.edu.au/pub/debian-cd/12.5.0/amd64/iso-cd/debian-12.5.0-amd64-netinst.iso",
                    headers={"Range": f"bytes=0-{1024*1024*10}"})
        logging.info("After get")
        return await r

@app.get("/")
async def main():
    logging.info("Start main/return")
    r = await async_request()
    return StreamingResponse(r.content)
```

### Description

I would expect the above code to return 10 megabytes of the file, hopefully streaming the request to the response directly
I have tested this code on Python 3.12.2 on Windows 11 and 3.11.2 on Debian 12.5

Instead, the response starts successfully, but stops after 327 bytes

Please forgive the contrived example, this works to demonstrate the issue

```
$ curl -sqN http://localhost:8000 --output - | grep -a Operating
f`f1f{f{fRfPSjjf6{6A{dfaOperating system load error.
```

(curl hangs until timeout)

Sidenote: this doesn't work at all in httpx, which can't produce a stream response that either asyncio or FastAPI can cope with. I might open another discussion about that one

### Operating System

Linux, Windows

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.11.2 / 3.12.2

**Comment by voltagex** [comment]
...this should probably be an issue, shouldn't it?

**Reply by YuriiMotov** [reply, marked as answer]
When you do `return await r` in `async_request`, you return StreamRead object, but at the same time you close client session, that is used by that StreamRead object (when you call `return` you exit from context manager `async with aiohttp.ClientSession() as client:`)

So, you need to turn your `async_request` into async generator and iterate through `r.content` inside:
```
import logging
import sys

import aiohttp
from fastapi import FastAPI
from fastapi.responses import Response, StreamingResponse

logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)

app = FastAPI()

async def async_request():
    async with aiohttp.ClientSession() as client:
        logging.info("Start get")
        r = await client.get(url="https://mirror.aarnet.edu.au/pub/debian-cd/12.5.0/amd64/iso-cd/debian-12.5.0-amd64-netinst.iso",
                    headers={"Range": f"bytes=0-{1024*1024*10}"})
        logging.info("After get")
        async for chunk in r.content:
            logging.info("next chunk")
            yield chunk

@app.get("/")
async def main():
    logging.info("Start main/return")
    return StreamingResponse(async_request())
```

**Reply by voltagex** [reply]
Thank you for this - I guess I'm a little too familiar with async/await in C#

Interesting note though, leaving the "next chunk" logging message causes the transfer rate to drop from ~10 megabytes/second to 1 megabyte/second!
